editor/uid_upgrade_tool.cpp

```diff
@@ -0,0 +1,120 @@
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "uid_upgrade_tool.h"
+
+#include "editor/editor_file_system.h"
+#include "editor/editor_log.h"
+#include "editor/editor_node.h"
+#include "editor/editor_settings.h"
+#include "editor/gui/editor_scene_tabs.h"
+#include "editor/gui/editor_toaster.h"
+#include "editor/themes/editor_scale.h"
+#include "servers/rendering_server.h"
+
+UIDUpgradeTool *UIDUpgradeTool::singleton = nullptr;
+
+void UIDUpgradeTool::_add_files(EditorFileSystemDirectory *p_dir, Vector<String> &r_resave_paths) {
+	for (int i = 0; i < p_dir->get_subdir_count(); i++) {
+		_add_files(p_dir->get_subdir(i), r_resave_paths);
+	}
+
+	for (int i = 0; i < p_dir->get_file_count(); i++) {
+		if (p_dir->get_file_type(i) == "PackedScene" || p_dir->get_file_type(i) == "Resource") {
+			r_resave_paths.append(p_dir->get_file_path(i));
+		}
+	}
+}
+
+void UIDUpgradeTool::prepare_upgrade() {
+	EditorSettings::get_singleton()->set_project_metadata("uid_upgrade_tool", "run_on_restart", true);
+	Vector<String> resave_paths;
+	_add_files(EditorFileSystem::get_singleton()->get_filesystem(), resave_paths);
+
+	EditorSettings::get_singleton()->set_project_metadata("uid_upgrade_tool", "resave_paths", resave_paths);
+
+	// Delay to avoid deadlocks, since this dialog can be triggered by loading a scene.
+	callable_mp(EditorNode::get_singleton(), &EditorNode::restart_editor).call_deferred(false);
+}
+
+void UIDUpgradeTool::begin_upgrade() {
+	EditorSettings::get_singleton()->set_project_metadata("uid_upgrade_tool", "run_on_restart", false);
+}
+
+void UIDUpgradeTool::finish_upgrade() {
+	EditorNode::get_singleton()->trigger_menu_option(EditorSceneTabs::SCENE_CLOSE_ALL, true);
+
+	Vector<String> resave_paths = EditorSettings::get_singleton()->get_project_metadata("uid_upgrade_tool", "resave_paths", Vector<String>());
```

Comment:
```suggestion
	Vector<String> resave_paths = EditorSettings::get_singleton()->get_project_metadata("uid_upgrade_tool", "resave_paths", Variant());
```
No reason to store empty array.